auction-server/src/simulator/main.rs

```diff
@@ -0,0 +1,191 @@
+use {
+    litesvm::{
+        types::{
+            FailedTransactionMetadata,
+            SimulatedTransactionInfo,
+        },
+        LiteSVM,
+    },
+    solana_client::{
+        rpc_config::RpcSendTransactionConfig,
+        rpc_response::{
+            Response,
+            RpcResult,
+        },
+    },
+    solana_rpc_client::nonblocking::rpc_client::RpcClient,
+    solana_sdk::{
+        account::{
+            Account,
+            ReadableAccount,
+        },
+        account_utils::StateMut,
+        bpf_loader_upgradeable::UpgradeableLoaderState,
+        commitment_config::CommitmentConfig,
+        pubkey::Pubkey,
+        signature::Signature,
+        transaction::VersionedTransaction,
+    },
+    std::{
+        collections::HashMap,
+        time::Instant,
+    },
+    tokio::sync::RwLock,
+};
+
+pub struct Simulator {
+    sender:      RpcClient,
+    receiver:    RpcClient,
+    pending_txs: RwLock<Vec<(VersionedTransaction, Instant)>>,
+}
+
+struct AccountsConfig {
+    accounts: HashMap<Pubkey, Account>,
+    programs: HashMap<Pubkey, Account>,
+}
+
+impl AccountsConfig {
+    fn new() -> Self {
+        Self {
+            accounts: Default::default(),
+            programs: Default::default(),
+        }
+    }
+
+    fn apply(&self, svm: &mut LiteSVM) {
+        for (key, account) in self.accounts.iter() {
+            svm.set_account(*key, account.clone()).unwrap();
+        }
+        for (key, account) in self.programs.iter() {
+            svm.add_program(
+                *key,
+                &(account.data()[UpgradeableLoaderState::size_of_programdata_metadata()..]),
+            );
+        }
+    }
+}
+
+impl Simulator {
+    pub fn new(sender: RpcClient, receiver: RpcClient) -> Self {
+        Self {
+            sender,
+            receiver,
+            pending_txs: Default::default(),
+        }
+    }
+
+    pub async fn fetch_pending_and_remove_old_txs(&self) -> Vec<VersionedTransaction> {
+        let now = Instant::now();
+        let mut pending_txs = self.pending_txs.write().await;
+        pending_txs.retain(|(_, time)| now.duration_since(*time).as_secs() < 5);
```

Comment:
is the 5 seconds here a proxy for whether the state change has been processed on-chain? why not just check the commitment statuses of the transactions? that may be easier and more robust via `getSignatureStatuses`

Reply:
yes, once we have a definite status we can remove it from this array. This will probably be handled by a separate thread that's either polling or subscribing